bftcosi/bftcosi.go

```diff
@@ -1,798 +1,239 @@
-
-	// wait until we have enough RoundPrepare commitments or timeout
-	// should do nothing if `c` is closed
-	if err := bft.readCommitChan(c, RoundPrepare); err != nil {
-		return err
-	}
 
-	// TODO this will not always work for non-star graphs
-	if len(bft.tempPrepareCommit) < len(bft.Children())-bft.allowedExceptions {
-		bft.signRefusal = true
-		log.Error("not enough prepare commitment messages")
-	}
-
-	commitment := bft.prepare.Commit(bft.Suite().RandomStream(), bft.tempPrepareCommit)
-	if bft.IsRoot() {
-		return bft.startChallenge(RoundPrepare)
-	}
-	return bft.SendToParent(&Commitment{
-		TYPE:       RoundPrepare,
-		Commitment: commitment,
-	})
+	return nil
 }
 
-// handleCommitmentCommit is similar to handleCommitmentPrepare except it is for
-// the commit phase.
-func (bft *ProtocolBFTCoSi) handleCommitmentCommit(c chan commitChan) error {
-	bft.tmpMutex.Lock()
-	defer bft.tmpMutex.Unlock() // NOTE potentially locked for the whole timeout
-
-	// wait until we have enough RoundCommit commitments or timeout
-	// should do nothing if `c` is closed
-	bft.readCommitChan(c, RoundCommit)
-
-	// TODO this will not always work for non-star graphs
-	if len(bft.tempCommitCommit) < len(bft.Children())-bft.allowedExceptions {
-		bft.signRefusal = true
-		log.Error("not enough commit commitment messages")
-	}
-
-	commitment := bft.commit.Commit(bft.Suite().RandomStream(), bft.tempCommitCommit)
-	if bft.IsRoot() {
-		// do nothing:
-		// stop the processing of the round, wait the end of
-		// the "prepare" round: calls startChallengeCommit
-		return nil
+func (bft *ProtocolBFTCoSi) initCosiProtocol(phase phase) (*protocol.CoSiRootNode, error) {
+	var name string
+	if phase == phasePrep {
+		name = bft.prepCosiProtoName
+	} else if phase == phaseCommit {
+		name = bft.commitCosiProtoName
+	} else {
+		return nil, fmt.Errorf("invalid phase %v", phase)
 	}
-	return bft.SendToParent(&Commitment{
-		TYPE:       RoundCommit,
-		Commitment: commitment,
-	})
-}
 
-// handleChallengePrepare collects the challenge-messages
-func (bft *ProtocolBFTCoSi) handleChallengePrepare(msg challengePrepareChan) error {
-	if bft.isClosing() {
-		return nil
-	}
-	ch := msg.ChallengePrepare
-	if !bft.IsRoot() {
-		bft.Msg = ch.Msg
-		bft.Data = ch.Data
-		// start the verification of the message
-		// acknowledge the challenge and send it down
-		bft.prepare.Challenge(ch.Challenge)
-	}
-	go func() {
-		bft.verifyChan <- bft.VerificationFunction(bft.Msg, bft.Data)
-	}()
-	if bft.IsLeaf() {
-		return bft.startResponse(RoundPrepare)
+	pi, err := bft.CreateProtocol(name, bft.Tree())
+	if err != nil {
+		return nil, err
 	}
-	return bft.sendToChildren(&ch)
-}
+	cosiProto := pi.(*protocol.CoSiRootNode)
+	cosiProto.CreateProtocol = bft.CreateProtocol
+	// NOTE We set it to n / 10 to have every sub-leader manage 10 nodes.
+	// This setting is bad if there are thousands of nodes as the root
+	// would need to manage hundres of sub-leaders.
```

Comment:
hundreds

Reply:
Not asking you to do it now, but if we need to make this adjustable later, will it be possible?

Reply:
Yes, we can add a `NSubtrees` property in bftcosi so that it is configurable. The value is propagated down to cosi and sub-cosi, so nothing else needs to be changed. Actually, there is a better way to compute `NSubtrees`, I'm updating the code.